fitbenchmarking/fitbenchmark_one_problem.py

```diff
@@ -85,47 +93,64 @@ def fitbm_one_prob(user_input, problem):
     return results_fit_problem
 
 
-def benchmark(controller, minimizers):
+def benchmark(controller, minimizers, num_runs):
     """
     Fit benchmark one problem, with one function definition and all
     the selected minimizers, using the chosen fitting software.
 
-    @param controller :: The software controller for the fitting
+    :param controller :: The software controller for the fitting
+    :type controller :: Object derived from BaseSoftwareController
     @param minimizers :: array of minimizers used in fitting
+    :type minimizers :: list
+    @param num_runs :: number of times controller.fit() is run to
+                       generate an average runtime
+    :type num_runs :: int
+
 
-    @returns :: nested array of result objects, per minimizer
-                and data object for the best fit data
+    :return :: tuple(results_problem, best_fit) nested array of
+               result objects, per minimizer and data object for
+               the best fit data
+    :rtype :: (list of FittingResult, plot_helper.data instance)
     """
     min_chi_sq, best_fit = None, None
     results_problem = []
 
     for minimizer in minimizers:
         controller.minimizer = minimizer
 
-        controller.prepare()
-
-        init_function_def = controller.problem.get_function_def(params=controller.initial_params,
-                                                                function_id=controller.function_id)
+        init_function_def = controller.problem.get_function_def(
+            params=controller.initial_params,
+            function_id=controller.function_id)
         try:
-            start_time = time.time()
-            controller.fit()
-            end_time = time.time()
-        except Exception as e:
-            print(e.message)
+            # Calls timeit repeat with repeat = num_runs and number = 1
+            runtime_list = \
+                timeit.Timer(setup=controller.prepare,
+                             stmt=controller.fit).repeat(num_runs, 1)
+            runtime = sum(runtime_list) / num_runs
+
+        # Catching all exceptions as this means runtime cannot be calculated
+        # pylint: disable=broad-except
+        except Exception as excp:
+            print(str(excp))
             controller.success = False
-            end_time = np.inf
-
-        runtime = end_time - start_time
+            runtime = np.inf
 
         controller.cleanup()
 
-        fin_function_def = controller.problem.get_function_def(params=controller.final_params,
-                                                               function_id=controller.function_id)
+        fin_function_def = controller.problem.get_function_def(
+            params=controller.final_params,
+            function_id=controller.function_id)
 
         if not controller.success:
             chi_sq = np.nan
             status = 'failed'
         else:
+            cv = np.std(runtime_list) / np.mean(runtime_list)
```

Comment:
Apologies, as my original suggestion was a bad metric for determining whether caching has occurred.  It would be better to test the ratio of the max to the min (as if a large number of runs were cached then the stdefv could be low despite a single outlier).  I've had a look at the timeit source code for Cython and the ratio they use is 4, so we could pick this to be consistent.

Also I think rather than raising the Warning maybe we should use `warnings.warn`, so just:
```
warnings.warn('The ratio of the max time to the min is larger than {}, which may indicate that caching has occurred in the timing results`.format(tolerance))
```

Reply:
Updated